This patch adds the sending of the -4 or -6 option to ssh when the
-4 or -6 option was given to rsync.

To use this patch, run these commands for a successful build:

    patch -p1 <patches/ssh-4-6-options.diff
    ./configure                         (optional if already run)
    make

based-on: a3377921ebe651cb7d2b969853cb9fe0e135ff75
diff --git a/main.c b/main.c
--- a/main.c
+++ b/main.c
@@ -80,6 +80,7 @@ extern BOOL want_progress_now;
 extern BOOL shutting_down;
 extern int backup_dir_len;
 extern int basis_dir_cnt;
+extern int default_af_hint;
 extern struct stats stats;
 extern char *stdout_format;
 extern char *logfile_format;
@@ -547,6 +548,11 @@ static pid_t do_cmd(char *cmd, char *machine, char *user, char **remote_argv, in
 			*t++ = '\0';
 		}
 
+		if ((t = strrchr(cmd, '/')) != NULL)
+		    t++;
+		else
+		    t = cmd;
+
 		/* check to see if we've already been given '-l user' in
 		 * the remote-shell command */
 		for (i = 0; i < argc-1; i++) {
@@ -566,20 +572,25 @@ static pid_t do_cmd(char *cmd, char *machine, char *user, char **remote_argv, in
 			args[argc++] = "-l";
 			args[argc++] = user;
 		}
+#ifdef AF_INET
+		if (default_af_hint == AF_INET && strcmp(cmd, "ssh") == 0) {
+			/* we're using ssh so we can add a -4 option */
+			args[argc++] = "-4";
+		}
+#endif
+#ifdef AF_INET6
+		if (default_af_hint == AF_INET6 && strcmp(cmd, "ssh") == 0) {
+			/* we're using ssh so we can add a -6 option */
+			args[argc++] = "-6";
+		}
+#endif
 		args[argc++] = machine;
 #endif
 
 		args[argc++] = rsync_path;
 
-		if (blocking_io < 0) {
-			char *cp;
-			if ((cp = strrchr(cmd, '/')) != NULL)
-				cp++;
-			else
-				cp = cmd;
-			if (strcmp(cp, "rsh") == 0 || strcmp(cp, "remsh") == 0)
-				blocking_io = 1;
-		}
+		if (blocking_io < 0 && (strcmp(t, "rsh") == 0 || strcmp(t, "remsh") == 0))
+			blocking_io = 1;
 
 		server_options(args,&argc);
 
diff --git a/rsync.1.md b/rsync.1.md
--- a/rsync.1.md
+++ b/rsync.1.md
@@ -3052,14 +3052,20 @@ your home directory (remove the '=' for that).
 
 0. `--ipv4`, `-4` or `--ipv6`, `-6`
 
-    Tells rsync to prefer IPv4/IPv6 when creating sockets.  This only affects
+    Tells rsync to prefer IPv4/IPv6 when creating sockets.  This affects
     sockets that rsync has direct control over, such as the outgoing socket
-    when directly contacting an rsync daemon.  See also these options in the
+    when directly contacting an rsync daemon, as well as forwaring the `-4` or
+    `-6` option to ssh, if rsync deduces that ssh is being used as the remote
+    shell.  For other remote shells, see the `--rsh` option.
+
+    See also these options in the
     `--daemon` mode section.
 
     If rsync was complied without support for IPv6, the `--ipv6` option will
     have no effect.  The `--version` output will tell you if this is the case.
 
+    See also these options in the `--daemon` mode section.
+
 0. `--checksum-seed=NUM`
 
     Set the checksum seed to the integer NUM.  This 4 byte checksum seed is
